_posts/2016-05-19-qcon-london.md

```diff
@@ -0,0 +1,162 @@
+* A slower data source will take longer to update itself, so at any given point in time the data sources might not be in sync, but they will be. (This is what’s known as eventual consistency). This might not make it suitable for situations where you present data from two or more data sources next to each other on a web page for example.
+* A slower data source won’t slow down updates being available in the others, in the way that transaction management would.
+* Since the reliability and consistency guarantees come from the single-threaded sequential nature of the consumers, scaling is achieved by sharding the events into separate logs (Kafka topics).
+* It leads to an architecture which can continue to grow without incurring large-scale technical debt, due to its loosely coupled nature. In this way, it goes nicely hand in hand with microservices should you wish to go down that route.
+
+It’s worth noting that financial systems, famous for stringent requirements around consistency, reliability and performance, very often use this kind of model.
+
+## Failure Testing
+
+So talking of reliability, the last talk I wanted to touch on is a hugely entertaining one I watched about **fault injection testing at Netflix**.
+In a pretty slick double act, academic [Peter Alvaro](https://twitter.com/palvaro) and [Kolton Andrus](https://twitter.com/koltonandrus) previously at Netflix, described how they met and collaborated on finding a way of discovering fault-intolerance in Netflix’s architectural estate: [Monkeys in Lab Coats: Applying Failure Testing Research at Netflix](http://www.infoq.com/presentations/failure-test-research-netflix).
+
+I highly recommend watching this talk, but the core concept was this:
+
+If you run 100 services and you’re interested in scenarios involving a _single service failure_, that’s quite easy to write fault injection tests for - there’s only 100 of them you need to run. But... it's probably not very interesting, because you’ll have built redundancy in (right?), so you won’t be discovering the “deep” bugs that happen when multiple things fail.
+
+So how big is the space of possible failures? It’s the power set of 100, which is **2^100**, or roughly **1,000,000,000,000,000,000,000,000,000,000**. That’s probably not going to be a viable number of test executions. [If each execution took 1 second, it would take longer than [the age of the universe](https://en.wikipedia.org/wiki/Age_of_the_universe) to run them all. About 40 trillion times longer in fact!]
+
+![What could possibly go wrong?](/images/whatcouldpossiblygowrong.png)
+
+Even if you decided to only look at combinations of 7 faults, that would mean a search space of **16 billion** [over 500 years of continuous executions each one second in duration]. Even just combinations of 4 would be **3 million** executions [maybe do-able in a month. But how much stuff changes in that month? And you’re still not testing for very deep bugs].
```

Comment:
Suggest changing to:

Even if you decided to only look at combinations of seven faults, that would mean a search space of **16 billion** (over 500 years of continuous executions each one second in duration); even just combinations of four would be **3 million** executions &emdash; this might be do-able in a month, but how much stuff changes in that month? Even then, you’re still not testing for very deep bugs.